File: app/src/androidTest/java/de/gaffga/android/zazentimer/utils/MeditationServiceIdlingResource.java
package de.gaffga.android.zazentimer.utils;

import androidx.test.espresso.IdlingResource;

public class MeditationServiceIdlingResource implements IdlingResource {

    private volatile ResourceCallback resourceCallback;
    private volatile boolean isIdle = true;

    @Override
    public String getName() {
        return "MeditationServiceIdlingResource";
    }

    @Override
    public boolean isIdleNow() {
        return isIdle;
    }

    @Override
    public void registerIdleTransitionCallback(ResourceCallback callback) {
        this.resourceCallback = callback;
        callback.onTransitionToIdle();
    }

    public void setBusy() {
        isIdle = false;
    }

    public void setIdle() {
        isIdle = true;
        notifyIdle();
    }

    private void notifyIdle() {
        if (resourceCallback != null) {
            resourceCallback.onTransitionToIdle();
        }
    }
}

File: app/src/androidTest/kotlin/at/priv/graf/zazentimer/utils/MeditationServiceIdlingResource.kt
package at.priv.graf.zazentimer.utils

import androidx.test.espresso.IdlingResource

class MeditationServiceIdlingResource : IdlingResource {

    @Volatile
    private var resourceCallback: IdlingResource.ResourceCallback? = null
    @Volatile
    private var isIdle = true

    override fun getName(): String = "MeditationServiceIdlingResource"

    override fun isIdleNow(): Boolean = isIdle

    override fun registerIdleTransitionCallback(callback: IdlingResource.ResourceCallback) {
        resourceCallback = callback
        callback.onTransitionToIdle()
    }

    fun setBusy() {
        isIdle = false
    }

    fun setIdle() {
        isIdle = true
        notifyIdle()
    }

    private fun notifyIdle() {
        resourceCallback?.onTransitionToIdle()
    }
}
